packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -17,81 +19,231 @@ import {
 import {and, pathStartsWith} from '../utils'
 
 type Client = ApiClients['shopperCustomers']
+type Customer = ShopperCustomersTypes.Customer
+type CustomerProductList = ShopperCustomersTypes.CustomerProductList
+type CustomerProductListResult = ShopperCustomersTypes.CustomerProductListResult
 
 const noop = () => ({})
 
-/** Invalidates the customer endpoint, but not derivative endpoints. */
-const invalidateCustomer = (parameters: Tail<QueryKeys['getCustomer']>): CacheUpdate => ({
-    invalidate: [{queryKey: getCustomer.queryKey(parameters)}]
-})
+// ReactQuery type def for the updater function uses type `unknown` for the first param,
+// this makes it hard to define the `createUpdateFunction` below, so we created this type.
+type Updater<T> = (data: T | undefined) => T | undefined
+
+// This is a slight variation of above, but we know the data is not null.
+type DataProcessor<T> = (data: NonNullable<T>) => T
+
+/**
+ * Create an update handler for the cache matrix. This updater will implicitly clone and assert
+ * that the data being passed to is is not undefined.
+ *
+ * @param update
+ * @returns
+ */
+const createUpdateFunction =
+    <T>(update: DataProcessor<T>): Updater<T> =>
+    (data) =>
+        !data ? data : update(clone(data))
```

Comment:
[nit] `!data ? data : ...` makes me think "what is `data`?" more than I want to
```suggestion
        data ? update(clone(data)) : undefined
```
This makes it more explicit what the return value is.

```suggestion
        data && update(clone(data))
```
Or this, if it doesn't cause a type/lint error.